File: convert.rb
# convert.rb
require 'mini_magick'
require 'find'
require 'optparse'

def convert_png_to_jpg(input_path, output_path, quality, resize_w, resize_h)
  image = MiniMagick::Image.open(input_path)
  # Конвертация в JPG (автоматически удаляет альфа-канал)
  image.format 'jpg'
  image.quality quality
  if resize_w > 0 && resize_h > 0
    image.resize "#{resize_w}x#{resize_h}"
  end
  image.write output_path
end

def process_files(inputs, quality, resize_w, resize_h, output_dir, overwrite, recursive)
  files = []
  inputs.each do |item|
    if File.file?(item) && item.downcase.end_with?('.png')
      files << item
    elsif File.directory?(item)
      if recursive
        Find.find(item) do |path|
          files << path if File.file?(path) && path.downcase.end_with?('.png')
        end
      else
        Dir.glob(File.join(item, '*.png')).each { |f| files << f }
      end
    elsif item.include?('*')
      Dir.glob(item).each { |f| files << f if f.downcase.end_with?('.png') }
    end
  end
  if files.empty?
    puts "Не найдено PNG-файлов."
    return
  end
  Dir.mkdir(output_dir) unless Dir.exist?(output_dir)
  total = files.size
  puts "Найдено #{total} PNG-файлов."
  files.each_with_index do |input_file, idx|
    out_name = File.basename(input_file, '.png') + '.jpg'
    out_path = File.join(output_dir, out_name)
    if File.exist?(out_path) && !overwrite
      puts "[#{idx+1}/#{total}] #{out_path} уже существует, пропуск."
      next
    end
    puts "[#{idx+1}/#{total}] Конвертация #{input_file} -> #{out_path}"
    begin
      convert_png_to_jpg(input_file, out_path, quality, resize_w, resize_h)
    rescue => e
      puts "  Ошибка при конвертации #{input_file}: #{e.message}"
    end
  end
  puts "Готово!"
end

options = {}
OptionParser.new do |opts|
  opts.banner = "Использование: ruby convert.rb <PNG-файлы/папки> [опции]"
  opts.on("--quality N", Integer, "Качество JPG (1-100)") { |v| options[:quality] = v }
  opts.on("--resize ШxВ", String, "Изменение размера") { |v| options[:resize] = v }
  opts.on("--output DIR", String, "Папка для сохранения") { |v| options[:output] = v }
  opts.on("--overwrite", "Перезаписывать") { options[:overwrite] = true }
  opts.on("--recursive", "Рекурсивный обход") { options[:recursive] = true }
end.parse!

quality = options[:quality] || 85
resize = options[:resize] ? options[:resize].split('x').map(&:to_i) : [0,0]
output_dir = options[:output] || '.'
overwrite = options[:overwrite] || false
recursive = options[:recursive] || false
inputs = ARGV

if inputs.empty?
  puts "Не указаны файлы или папки."
  exit
end

process_files(inputs, quality, resize[0], resize[1], output_dir, overwrite, recursive)

File: convert.cs
// convert.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using SixLabors.ImageSharp;
using SixLabors.ImageSharp.PixelFormats;
using SixLabors.ImageSharp.Processing;
using SixLabors.ImageSharp.Formats.Jpeg;

class Convert
{
    static void ConvertPngToJpg(string inputPath, string outputPath, int quality, int resizeW, int resizeH)
    {
        using (var image = Image.Load<Rgba32>(inputPath))
        {
            // Конвертация в RGB (JPG не поддерживает альфа)
            var rgbImage = image.CloneAs<Rgb24>();
            // Изменение размера
            if (resizeW > 0 && resizeH > 0)
            {
                rgbImage.Mutate(x => x.Resize(resizeW, resizeH));
            }
            // Сохранение
            var encoder = new JpegEncoder
            {
                Quality = quality,
                ColorType = JpegColorType.YCbCr,
                ChromaSubsampling = JpegChromaSubsampling.Ratio420
            };
            Directory.CreateDirectory(Path.GetDirectoryName(outputPath));
            rgbImage.Save(outputPath, encoder);
        }
    }

    static void ProcessFiles(List<string> inputs, int quality, int resizeW, int resizeH, string outputDir, bool overwrite, bool recursive)
    {
        var files = new List<string>();
        foreach (var item in inputs)
        {
            if (File.Exists(item) && item.ToLower().EndsWith(".png"))
            {
                files.Add(item);
            }
            else if (Directory.Exists(item))
            {
                var search = recursive ? SearchOption.AllDirectories : SearchOption.TopDirectoryOnly;
                files.AddRange(Directory.GetFiles(item, "*.png", search));
            }
            else if (item.Contains("*"))
            {
                // упрощённо — маски не обрабатываем
            }
        }
        if (files.Count == 0)
        {
            Console.WriteLine("Не найдено PNG-файлов.");
            return;
        }
        Directory.CreateDirectory(outputDir);
        int total = files.Count;
        Console.WriteLine($"Найдено {total} PNG-файлов.");
        for (int i=0; i<total; i++)
        {
            var inputFile = files[i];
            var outName = Path.GetFileNameWithoutExtension(inputFile) + ".jpg";
            var outPath = Path.Combine(outputDir, outName);
            if (File.Exists(outPath) && !overwrite)
            {
                Console.WriteLine($"[{i+1}/{total}] {outPath} уже существует, пропуск.");
                continue;
            }
            Console.WriteLine($"[{i+1}/{total}] Конвертация {inputFile} -> {outPath}");
            try
            {
                ConvertPngToJpg(inputFile, outPath, quality, resizeW, resizeH);
            }
            catch (Exception e)
            {
                Console.WriteLine($"  Ошибка при конвертации {inputFile}: {e.Message}");
            }
        }
        Console.WriteLine("Готово!");
    }

    static void Main(string[] args)
    {
        if (args.Length == 0)
        {
            Console.WriteLine("Использование: dotnet run <PNG-файлы/папки> [--quality N] [--resize ШxВ] [--output DIR] [--overwrite] [--recursive]");
            return;
        }
        var inputs = new List<string>();
        int quality = 85;
        int resizeW = 0, resizeH = 0;
        string outputDir = ".";
        bool overwrite = false;
        bool recursive = false;
        for (int i=0; i<args.Length; i++)
        {
            switch (args[i])
            {
                case "--quality":
                    if (i+1 < args.Length) quality = int.Parse(args[++i]);
                    break;
                case "--resize":
                    if (i+1 < args.Length)
                    {
                        var s = args[++i];
                        var parts = s.Split('x');
                        if (parts.Length == 2)
                        {
                            resizeW = int.Parse(parts[0]);
                            resizeH = int.Parse(parts[1]);
                        }
                    }
                    break;
                case "--output":
                    if (i+1 < args.Length) outputDir = args[++i];
                    break;
                case "--overwrite":
                    overwrite = true;
                    break;
                case "--recursive":
                    recursive = true;
                    break;
                default:
                    inputs.Add(args[i]);
                    break;
            }
        }
        ProcessFiles(inputs, quality, resizeW, resizeH, outputDir, overwrite, recursive);
    }
}
